sdk/trace/span.go

```diff
@@ -140,17 +207,24 @@ func (s *span) End(options ...trace.SpanOption) {
 	}
 	config := trace.NewSpanConfig(options...)
 	s.endOnce.Do(func() {
+		s.mu.Lock()
+		if config.Timestamp.IsZero() {
+			// TODO: Does it make sense to store the end time at the very
+			// beginning of span.End() to minimize the impact of the operations
+			// above on the span's duration? Doing many operations prior to
+			// marking the end of the span could become significant when
+			// instrumenting very quick application operations.
```

Comment:
If setting the end time were moved up it would be necessary to also move up the `if !s.IsRecording()` check above (or remove it entirely, perhaps).

Reply:
Makes sense. I'll take a look at this. I'll try to check if storing the end time early has any impact at all.

Reply:
I've tried sleeping inside `End()` before setting `s.endTime` and as expected this makes the span's duration "artificially" long. As far as I can tell we want to store the end timestamp ASAP because the thing we are tracing is the **user** operation, not the time it takes us to finish a span (right?).

>If setting the end time were moved up it would be necessary to also move up the if !s.IsRecording() check above (or remove it entirely, perhaps).

Actually @Aneurysm9 I don't have to touch `IsRecording()` because I only store the end time in a new variable at the top, and then set it in `span.endTime` only if we get to this part of the function (i.e. if we didn't return earlier).

I have another problem though: If `End()` gets called more than once, my current logic overwrites the original end time which shouldn't happen. I'm looking into this.

Reply:
Looks like we might be using `s.endOnce` unnecessarily because if `End()` is called a 2nd time, `IsRecording()` returns `false` and we correctly return without trying to overwrite the end time. I'll open a new PR to refactor `End()` as this is starting to be out of scope here IMO.

Reply:
>I'll open a new PR to refactor End() as this is starting to be out of scope here IMO.

...or not :see_no_evil: I don't have the right interfaces for refactoring properly until this is merged. Oh well... Let me see what I come up with.

Reply:
I've pushed a new commit which refactors `span.End()`. Looks like we can address the problem discussed here and also simplify the idempotence logic.